### ISSUE TITLE
discoverd: Heartbeating

### ISSUE THREAD

**Issue by benbjohnson** [flynn/flynn#2158]
## Overview

Currently `discoverd` can experience high write load because instances heartbeat their status every few seconds and this heartbeat then goes through raft consensus. Because it goes through consensus, all `discoverd` instances are kept in sync however this has a high cost.

Jonathan had mentioned moving heartbeats to a separate in-memory raft instance to avoid disk I/O, however, I don't think we need the strong consistency that raft provides. Also, raft probably won't do well to recover if a quorum of nodes loses all their data.

## Proposed Change

Instead of moving to another raft cluster, I propose that `discoverd` simply forwards heartbeats to the other nodes in the cluster. The state can be kept in-memory and is only used when a node is leader. If two separate heartbeats come in for the same instance, it is easy to merge them -- simply use the higher timestamp. In the event of a crash, the node simply waits for new heartbeats to come in before it can change leadership for a service.

We can piggyback onto the TCP connection that already exists by multiplexing it. Consul does this and I've written a TCP multiplexer before and it works well. In fact, we could also multiplex the HTTP traffic so it all goes over the same port.

**Comment by titanous** [comment]
This sounds reasonable to me. Just to be clear, non-leader nodes will still be able to serve reads, right?

**Comment by benbjohnson** [comment]
Yes, everything but last heartbeat will still go through consensus.

**Comment by josephglanville** [comment]
This seems pretty good.

On a new node joining or coming up it can simply wait a heartbeat interval or 2 before starting to service reads via the HTTP API which will be enough to bring it fully in-sync.
